### PR TITLE
Improve format list and downloads tab

### LINKED ISSUE #15: Option: Use stream in/for media player
In addition to downloading, it would be nice to have the option to stream the media file with the default media player or chosen one.

### LINKED ISSUE #16: Update the list of files when a download is removed
Great app. Please add a feature to delete downloads and a feature to update the list of videos if the file is manually deleted. 

As itis now manually deleted files remain in the list of videos 

Thank you

### LINKED ISSUE #29: Share button
can this app implement a share button for sharing videos in the downloads tab

### LINKED ISSUE #36: Add ability to clear download history
I'd really like an option to clear the download history from the downloads tab. 

Thanks.

### PR DESCRIPTION
**Pull request by xibr**
Improve format list and downloads tab

add option to share and delete  files in the downloads tab 
add share button from which can copy the link or open it in any other player such as [mpv android.](https://play.google.com/store/apps/details?id=is.xyz.mpv) 

Closes #36 
Closes #29 
Closes #16 
Closes #15 

[armeabi-v7a-debug.apk](https://github.com/xibr/xibr.github.io/raw/master/app-armeabi-v7a-debug.apk)

[x86-debug.apk](https://github.com/xibr/xibr.github.io/raw/master/app-x86-debug.apk)

### CODE REVIEW

**Comment by yausername** [comment]
I will test these changes tonight.

File: app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt [hunk 1]
```diff
@@ -14,6 +14,9 @@ interface DownloadsDao {
     @Delete
     suspend fun delete(item: Download)
 
+    @Query("DELETE from downloads_table WHERE downloaded_path = :filePath")
```
Comment (yausername) [inline app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt:17]: How about delete by `id` instead?

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -47,4 +47,7 @@
     <string name="nav_graph_label_downloads">Downloads</string>
     <string name="nav_graph_label_settings">Settings</string>
     <string name="nav_graph_label_ytdlp">yt-dlp</string>
+    <string name="share">share</string>
```
Comment (yausername) [inline app/src/main/res/values/strings.xml:50]:
```suggestion
    <string name="share">Share</string>
```

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -47,4 +47,7 @@
...
+    <string name="delete">delete</string>
```
Comment (yausername) [inline app/src/main/res/values/strings.xml:51]:
```suggestion
    <string name="delete">Delete</string>
```

File: app/src/main/java/org/yausername/dvd/adapters/DownloadsAdapter.kt [hunk 4]
```diff
@@ -77,4 +95,43 @@ class DownloadsAdapter : RecyclerView.Adapter<DownloadsAdapter.ViewHolder>() {
...
+    private fun startDelete(name: String, path: String, context: Context) {
+        val workTag = name
+        val workManager = WorkManager.getInstance(context.applicationContext!!)
+        val state = workManager.getWorkInfosByTag(workTag).get()?.getOrNull(0)?.state
+
+        if (state === WorkInfo.State.RUNNING || state === WorkInfo.State.ENQUEUED) {
+            return
+        }
+        val workData = workDataOf(
+            DeleteWorker.fileUriKey to path,
```
Comment (yausername) [inline app/src/main/java/org/yausername/dvd/adapters/DownloadsAdapter.kt:122]: how about we just send the `id` and use it to fetch the `Download` object in `DeleteWorker`

File: app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt (hunk 1 above)
Comment (yausername) [reply app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt:17]: there is already a delete function on line 15

**Comment by yausername** [comment]
The change looks good. It's a good improvement.
I left some minor comments about the delete flow.

File: app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt (hunk 1 above)
Comment (xibr) [reply app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt:17]: > How about delete by `id` instead?

in `DownloadsDao.kt`
``` diff
diff --git a/app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt b/app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt
index 29e8933..9ed2387 100644
--- a/app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt
+++ b/app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt
@@ -14,8 +14,8 @@ interface DownloadsDao {
     @Delete
     suspend fun delete(item: Download)
 
-    @Query("DELETE from downloads_table WHERE downloaded_path = :filePath")
-    fun deleteByUri(filePath: String)
+    @Query("DELETE from downloads_table WHERE id = :id")
+    fun deleteById(id: Long)
 
     @Query("SELECT * from downloads_table ORDER BY timestamp DESC")
     fun getAllDownloads(): LiveData<List<Download>>

```
and `DownloadsRepository.kt`
``` diff
diff --git a/app/src/main/java/org/yausername/dvd/database/DownloadsRepository.kt b/app/src/main/java/org/yausername/dvd/database/DownloadsRepository.kt
index 92b9206..9eb46f0 100644
--- a/app/src/main/java/org/yausername/dvd/database/DownloadsRepository.kt
+++ b/app/src/main/java/org/yausername/dvd/database/DownloadsRepository.kt
@@ -18,7 +18,8 @@ class DownloadsRepository(private val downloadsDao: DownloadsDao) {
         downloadsDao.delete(download)
     }
 
-    fun deleteByUri(filePath: String) {
-        downloadsDao.deleteByUri(filePath)
+    fun deleteById(id: Long) {
+        downloadsDao.deleteById(id)
     }
+
 }

```
> there is already a delete function on line 15

Yes but how do I use it ? I used the simple method.

Comment (yausername) [reply app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt:17]: > Yes but how do I use it ? I used the simple method.


```kotlin
val download = repository.getById(id)
repository.delete(download)
```

You will have to define a new function to `getById`
```
@Query("SELECT * from downloads_table WHERE id = :id")
fun getById(id: Long): Download
```

Comment (xibr) [reply app/src/main/java/org/yausername/dvd/database/DownloadsDao.kt:17]: Thanks.

**Review by yausername — APPROVED** [review]

**Pull request merged**